src/audio/module_adapter/module/cadence.c

```diff
@@ -207,16 +254,100 @@ static int cadence_codec_post_init(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
 static int cadence_codec_init(struct processing_module *mod)
 {
-	int ret;
+	const struct ipc4_cadence_module_cfg *cfg;
+	struct module_data *codec = &mod->priv;
+	struct cadence_codec_data *cd;
+	struct module_config *setup_cfg;
 	struct comp_dev *dev = mod->dev;
+	int ret;
+
+	comp_dbg(dev, "cadence_codec_init() start");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
+	if (!cd) {
+		comp_err(dev, "cadence_codec_init(): failed to allocate memory for cadence codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+	codec->mpd.init_done = 0;
+	cd->self = NULL;
+	cd->mem_tabs = NULL;
+	cd->api = NULL;
+	cd->setup_cfg.avail = false;
+
+	/* copy the setup config only for the first init */
+	if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
+		setup_cfg = &cd->setup_cfg;
+
+		cfg = (const struct ipc4_cadence_module_cfg *)codec->cfg.init_data;
+
+		/* allocate memory for set up config */
+		setup_cfg->data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  cfg->param_size);
+		if (!setup_cfg->data) {
+			comp_err(dev, "cadence_codec_init(): failed to alloc setup config");
+			ret = -ENOMEM;
+			goto free;
+		}
+
+		/* allocate memory for runtime set up config */
+		codec->cfg.data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  cfg->param_size);
+		if (!codec->cfg.data) {
+			comp_err(dev, "cadence_codec_init(): failed to alloc runtime setup config");
+			ret = -ENOMEM;
+			goto free_cfg;
+		}
+
+		codec->cfg.size = cfg->param_size;
+		ret = memcpy_s(codec->cfg.data, codec->cfg.size,
+			       cfg->param, cfg->param_size);
```

Comment:
you're overwriting the whole just allocated `codec->cfg.data`, why did you allocate it with `rzalloc()` then? Also, using different sizes for source and destination in `memcpy_s()` above is more misleading here than logical. You allocated `codec->cfg.data` with size `cfg->param_size` so, **that** is its size. I think it would look more logical to me as
```
tmp_size = cfg->param_size;
codec->cfg.data = rmalloc(..., tmp_size);
...
ret = memcpy_s(codec->cfg.data, tmp_size, cfg->param, cfg->param_size);
codec->cfg.size = tmp_size;
```

Reply:
There is no sense to use memcpy_s  with same argument passed for dest_size and count. For that purpose we have basic memcpy(). Safe version additionaly checks destination buffer size if data to be copied fits the buffer. In my opinion this is correct usage of _s function.

Reply:
@mbukowsk I disagree. Firstly just using an additional variable to *hide* the fact that you're using the same value doesn't suddenly make the code meaningful. Secondly `memcpy_s()` *still* makes sense when used with the same size - it verifies that the source and the destination don't overlap. Thirdly your version uses the same value just as much as mine. The difference is that in my version the same variable is at least used both for the allocation and copying, that way it is clear that that is the size of the buffer. Even though I'm personally not a fan of those "safe" versions, if we do use them, let's use them in the best possible way.